тесты к эндпоинту встреч

Cubaser opened the pull request:
Тесты для эндпойнтов problem_meetings.py https://github.com/Studio-Yandex-Practicum/Tabit/issues/213
Пофиксил post и patch запросы

Cubaser committed:
фикс форматирования

Cubaser committed:
фикс форматирования

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+    Тесты post запросов к эндпоинту meetings.
+    """
+
+    async def create_problem(self, async_session, company, owner, name='Test Problem'):
```
Создание записей в бд делаем в фикстурах. Смотри пример company_for_test() в conftest.py/

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        await async_session.refresh(problem)
+        return problem.id
+
+    def build_meeting_data(
```
Незачем в каждом тесте прописывать одинаковые значения для генерации данных для встреч. Пропиши по умолчанию рандомные данные внутри метода и при необходимости уже можешь обновлять поля для определенных тестов. 

Смотри пример метод generate_company_data в test_tabit_management_company.py

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+            members=members,
+        )
+        response_meeting = await client.post(
+            f'/api/v1/{company.slug}/problems/{problem_id}/meetings', json=meeting_data
```
URL запросов нужно перенести в константы. 

Пример, как сделать константы динамически можешь посмотреть тут: https://github.com/Studio-Yandex-Practicum/Tabit/pull/207#discussion_r1995537537

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        response_meeting = await client.post(
+            f'/api/v1/{company.slug}/problems/{problem_id}/meetings', json=meeting_data
+        )
+        assert response_meeting.status_code == 201
```
Помимо проверки статус кода, можно и нужно ещё проверять тело ответа, что соответствует тем значениям, которые мы отправляли в теле самого запроса

DinarM replied:
Номера статус кодов проставляем с помощью библиотеки from fastapi import status

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        variants = (
+            (problem_id, date_meeting, 201),
+            (second_problem_id, date_meeting, 400),
+            (second_problem_id, past_date_meeting, 422),
```
Тесты должны быть атомарными, позитивный и негативный кейс не может быть в одном тесте одновременно.

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+            (second_problem_id, date_meeting, 400),
+            (second_problem_id, past_date_meeting, 422),
+        )
+        for problem_id, date, status in variants:
```
Вместо цикла for, можно использовать параметризацию пайтеста(@pytest.mark.parametrize), для проверки нескольких значений одного поля

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+            response_meeting = await client.post(
+                f'/api/v1/{company.slug}/problems/{problem_id}/meetings', json=meeting_data
+            )
+            assert response_meeting.status_code == status
```
Помимо статус кода необходимо ещё проверить текст ошибки на соответствие

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        )
+
+        assert response.status_code == 200
+        # response_data = response.json()
+        # meeting = await async_session.get(Meeting, meeting_id)
+        # assert response_data['id'] == meeting_id
+        # assert response_data['title'] == meeting.title
+        # assert response_data['description'] == meeting.description
+        # assert response_data['date_meeting'] == meeting.date_meeting.strftime('%Y-%m-%d')
+        # assert response_data['status'] == meeting.status
+        # # assert response_data['place'] == meeting.place
+        # # assert response_data['created_at'] == meeting.created_at.isoformat()
+        # # assert response_data['updated_at'] == meeting.updated_at.isoformat()
```
?

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        owner = await employee_of_company()
+        company = await company_for_test()
+        problem_id = await self.create_problem(async_session, company, owner)
+        response = await client.get(f'/api/v1/{company.slug}/problems/{problem_id}/meetings')
+        assert response.status_code == 200
+        response_data = response.json()
+        assert isinstance(response_data, list)
+        assert len(response_data) == 0
```
Лучше разделить на два теста:
1) Получить пустой список, когда нет встреч
2) Добавить записи и проверить на кол-во

Так как у нас для каждого теста чистая база данных, делать предусловие нет смысла как мне кажется.

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        await async_session.refresh(problem)
+        return problem.id
+
+    async def create_meeting(self, async_session, problem_id, owner, title='Test Meeting'):
```
Переносим в фикстуры

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        owner = await employee_of_company()
+        company = await company_for_test()
+        meeting = await self.create_meeting(async_session, owner, company)
+        variants = (
```
Делаем два отдельных теста, позитивный и негативный

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+
+    async def create_problem(self, async_session, company, owner, name='Test Problem'):
+        """
+        Зосдание записи Problem в бд.
```
Орфографические ошибки:
	•	Зосдание → Создание
	•	со вмеми полями → со всеми полями
	•	Тест успешного удаление встречи. → Тест успешного удаления встречи.

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,493 @@
+        assert response.status_code == 200
+        response_data = response.json()
+        assert isinstance(response_data, list)
+        assert len(response_data) == 3
```
Добавь ещё проверку, что каждый элемент ответа содержит title, date_meeting, status и другие ключи

DinarM reviewed (commented):
Тесты в целом хорошие, но есть возможности для улучшения.

Cubaser committed:
Merge branch 'dev' of https://github.com/Studio-Yandex-Practicum/Tabit into feature/test_problem_meetings
Обновление ветки

Cubaser committed:
исправлены замечания PR

Cubaser committed:
исправлены замечания PR

AndreyZherdetskiy committed:
Merge branch 'dev' into feature/test_problem_meetings

Austin-PAVrus committed:
Добавлены тесты на эндпоинт /api/v1/auth/me (#236)

Добавлены тесты на эндпоинт /api/v1/auth/me для обычных пользователей и
модераторов компаний

Тикет вашего PR (если есть):
    Тесты для эндпойнтов company_user.py #210

---------

Co-authored-by: Konstantin Raikhert <69113745+KonstantinRaikhert@users.noreply.github.com>
Co-authored-by: AndreyZherdetskiy <138036940+AndreyZherdetskiy@users.noreply.github.com>

Austin-PAVrus committed:
Добавлены тесты на эндпоинт /api/v1/auth/me (#236)

Добавлены тесты на эндпоинт /api/v1/auth/me для обычных пользователей и
модераторов компаний

Тикет вашего PR (если есть):
    Тесты для эндпойнтов company_user.py #210

---------

Co-authored-by: Konstantin Raikhert <69113745+KonstantinRaikhert@users.noreply.github.com>
Co-authored-by: AndreyZherdetskiy <138036940+AndreyZherdetskiy@users.noreply.github.com>

PotashevIlya committed:
Исправлены баги в GET и POST запросах к эндпоинтам promlems (#243)

1. Исправлен баг, когда при GET запросе на список проблем возвращались
все проблемы из базы без привязки к определенной компании.
2. При отправке POST-запроса на создание проблемы поле "members" теперь
опционально. Возможны 3 сценария:
    а) Поле не передано вообще. 
    б) В поле передан пустой список. 
в) В поле передан список "members", состоящий из UUID пользователей
Tabit.
При этом создатель проблемы ("owner") автоматически становится
участником проблемы.
3. Поле "company_id" в модели Problem изменено на "company_slug". Также
данное поле теперь ссылается не на id компании, а на её слаг.
4. Поле "id" удалено из модели AssociationUserProblem, PK в данной
модели теперь состоит из полей "left_id"(ссылается на пользователя) и
"rigth_id"(ссылается на проблему).
5. В соответствии с изменениями в моделях изменён файл миграций.

---------

Co-authored-by: Ilya Potashev <ilya@DESKTOP-MTT1JD0>
Co-authored-by: AndreyZherdetskiy <138036940+AndreyZherdetskiy@users.noreply.github.com>

PotashevIlya committed:
Исправлены баги в GET и POST запросах к эндпоинтам promlems (#243)

1. Исправлен баг, когда при GET запросе на список проблем возвращались
все проблемы из базы без привязки к определенной компании.
2. При отправке POST-запроса на создание проблемы поле "members" теперь
опционально. Возможны 3 сценария:
    а) Поле не передано вообще. 
    б) В поле передан пустой список. 
в) В поле передан список "members", состоящий из UUID пользователей
Tabit.
При этом создатель проблемы ("owner") автоматически становится
участником проблемы.
3. Поле "company_id" в модели Problem изменено на "company_slug". Также
данное поле теперь ссылается не на id компании, а на её слаг.
4. Поле "id" удалено из модели AssociationUserProblem, PK в данной
модели теперь состоит из полей "left_id"(ссылается на пользователя) и
"rigth_id"(ссылается на проблему).
5. В соответствии с изменениями в моделях изменён файл миграций.

---------

Co-authored-by: Ilya Potashev <ilya@DESKTOP-MTT1JD0>
Co-authored-by: AndreyZherdetskiy <138036940+AndreyZherdetskiy@users.noreply.github.com>

DinarM commented on tests/conftest.py:
```diff
@@ -444,3 +445,47 @@ async def employee_refresh_token(get_token_for_user, employee):
+
+
+@pytest_asyncio.fixture
+async def problem_for_meeting(async_session: AsyncSession):
```
Для упрощения кода в тестах предлагаю добавить в фикстуры problem_for_meeting и create_meeting возможность создания объектов company и owner по умолчанию, если они не переданы в параметрах.

Это можно реализовать по аналогии с employee_of_company, где при необходимости автоматически создаётся компания.

Таким образом, в тестах не потребуется явно создавать переменные для этих объектов.

Текущий вариант:
``` 
        owner = await employee_of_company()
        company = await company_for_test()
        problem_id = await problem_for_meeting(company_id=company.id, owner_id=owner.id)
```
Предлагаемый вариант:
```
problem = await problem_for_meeting()
```
После чего необходимые ID компании и владельца можно получать из самого объекта problem.

Этот подход сделает код тестов чище и избавит от дублирования создания стандартных объектов.

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+        employee_of_company,
+        problem_for_meeting,
+        company_for_test,
+        create_meeting,
```
лишняя фикстура, которая не вызывается

Cubaser replied:
все вызываются

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+        employee_of_company,
+        problem_for_meeting,
+        company_for_test,
+        create_meeting,
```
лишняя фикстура, которая не вызывается

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+        response_data = response_meeting.json()
+        assert response_data['title'] == meeting_data['title']
+        assert response_data['date_meeting'] == meeting_data['date_meeting']
+        assert response_data['status'] == meeting_data['status']
```
предлагаю в этом тесте проверить наличие необязательных полей, а в test_create_meeting_with_required_fields проверку обязательных в ответе

DinarM commented on tests/conftest.py:
```diff
@@ -444,3 +445,47 @@ async def employee_refresh_token(get_token_for_user, employee):
+    async def func(problem_id, owner_id, count=1, date=None, data=False):
+        meetings_ids = []
+        meetings_data = []
+        for i in range(count):
```
цикл for предлагаю убрать и уже в самих тестах использовать этот цикл, если нужно создать несколько объектов, чтобы не усложнять фикстуру

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+    """
+
+    @pytest.mark.asyncio
+    async def test_delet_meeting(
```
Орфографическая ошибка:
test_delet_meeting → test_delete_meeting

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+        assert response.status_code == status.HTTP_204_NO_CONTENT
+
+    @pytest.mark.asyncio
+    async def test_delet_no_meeting(
```
Орфографическая ошибка:
test_delet_meeting → test_delete_meeting

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+
+
+def generate_meeting_data(
+    problem_id, owner_id, all_fields=False, members=None, meeting_title=None, meeting_date=None
```
А зачем нужны параметры meeting_title=None, meeting_date=None?

Ты можешь создать словарь generate_meeting_data с рандомными значениями и потом поменять любое значение.

пример из тестов лицензии:
```
data = generate_license_data()
data['license_term'] = 0
```

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+            fisrt_date,
+            second_date,
+        )
+        for date in variants:
```
Лучше применить pytest.mark.parametrize, чтобы было видно, какие именно параметры используются

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+        problem_id = await problem_for_meeting(company_id=company.id, owner_id=owner.id)
+        meetings_ids = await create_meeting(problem_id=problem_id, owner_id=owner.id)
+        meeting_id = meetings_ids[0]
+        variants = (
```
Можно упростить, заменив variants на pytest.mark.parametrize. В дев ветке уже залили пример с использованием параметризации

DinarM commented on tests/test_problem_meetings.py:
```diff
@@ -0,0 +1,422 @@
+            ),
+            json={'title': 'No Matter'},
+        )
+        assert response_patch.status_code == status.HTTP_404_NOT_FOUND
```
Сейчас проверяется только код ответа, но лучше добавить проверку response.json(), чтобы убедиться, что возвращается ожидаемое сообщение.

levisserena committed:
Bugfix/image base64 (#244)

При создании или изменении компании поле `logo` теперь принимает строку
в кодировке Base64.
Сохраняет файл в директорию media (должна быть создана при
развертывании, и должна быть `volume` в контейнере, для сбора статики),
распределяя в одноименную директорию `logo` (если не будет - создаст). В
БД в поле `logo` сохранится путь до файла формата:
```
media/logo/slug_company.png
```
Именуется файл по `slug` компани.
Поправил тесты. Поправил коллекцию Postman. Изменил CRUD у компаний.
П.С. Удалил пустые файлы в модулях.

levisserena committed:
Bugfix/image base64 (#244)

При создании или изменении компании поле `logo` теперь принимает строку
в кодировке Base64.
Сохраняет файл в директорию media (должна быть создана при
развертывании, и должна быть `volume` в контейнере, для сбора статики),
распределяя в одноименную директорию `logo` (если не будет - создаст). В
БД в поле `logo` сохранится путь до файла формата:
```
media/logo/slug_company.png
```
Именуется файл по `slug` компани.
Поправил тесты. Поправил коллекцию Postman. Изменил CRUD у компаний.
П.С. Удалил пустые файлы в модулях.

Cubaser committed:
Решены конфликты с dev

Cubaser committed:
Решены конфликты с dev

Cubaser committed:
Merge branch 'feature/test_problem_meetings' of https://github.com/Studio-Yandex-Practicum/Tabit into feature/test_problem_meetings
"Решены конфликты с dev"
